### ISSUE TITLE
Overload selection failure with protocols

### ISSUE THREAD

**Issue by bb010g** [microsoft/pyright#3212]
**Describe the bug**
Overload selection seems to break down with a certain pattern of protocols & higher-order functions, such that union members that should match 1-to-1 with overloads in a protocol suddenly try to match 1-to-many.

**To Reproduce**
See example code below. If you remove `@decorates_my_callable`, the example type-checks.

**Expected behavior**
I expect unions arms to consistently match various arms of an overloaded function.

**Screenshots or Code**
```python
from typing import Literal, Protocol, TypeVar, overload


Tv_val_0 = TypeVar('Tv_val_0', contravariant=True)
Tv_val_0_co = TypeVar('Tv_val_0_co', covariant=True)
Tv_val_1 = TypeVar('Tv_val_1', contravariant=True)
Tv_val_1_co = TypeVar('Tv_val_1_co', covariant=True)
Tv_result = TypeVar('Tv_result', covariant=True)
Tv_result_co = TypeVar('Tv_result_co', contravariant=True)
Tv_my_callable_0 = \
    TypeVar('Tv_my_callable_0', bound='MyCallable[object, object, object]')
Tv_my_callable_1 = \
    TypeVar('Tv_my_callable_1', bound='MyCallable[object, object, object]')


class My0Callable(Protocol[Tv_val_0, Tv_val_1, Tv_result]):
    def __call__(
        self, /, tag: Literal[0], arg_0: Tv_val_0, arg_1: Tv_val_1
    ) -> Tv_result: ...


class My1Callable(Protocol[Tv_val_0, Tv_val_1, Tv_result]):
    def __call__(
        self, /, tag: Literal[1], arg_0: Tv_val_1, arg_1: Tv_val_0
    ) -> Tv_result: ...


MyCallable = (
    My0Callable[Tv_val_0, Tv_val_1, Tv_result] |
    My1Callable[Tv_val_0, Tv_val_1, Tv_result]
)


class MyCallableDecorator(
    Protocol[
        Tv_val_0_co, Tv_val_1_co, Tv_result_co, Tv_val_0, Tv_val_1, Tv_result
    ]
):
    @overload
    def __call__(
        self, __callable_0: My0Callable[Tv_val_0_co, Tv_val_1_co, Tv_result_co]
    ) -> My0Callable[Tv_val_0, Tv_val_1, Tv_result]: ...

    @overload
    def __call__(
        self, __callable_1: My1Callable[Tv_val_0_co, Tv_val_1_co, Tv_result_co]
    ) -> My1Callable[Tv_val_0, Tv_val_1, Tv_result]: ...


def decorates_my_callable(
    __decorator: MyCallableDecorator[
        Tv_val_0_co, Tv_val_1_co, Tv_result_co, Tv_val_0, Tv_val_1, Tv_result
    ]
) -> MyCallableDecorator[
    Tv_val_0_co, Tv_val_1_co, Tv_result_co, Tv_val_0, Tv_val_1, Tv_result
]:
    return __decorator


def my_decorator(
    __my_callable_outer: Tv_my_callable_0
) -> Tv_my_callable_0:
    @decorates_my_callable
    def my_decorator_inner(
        __my_callable_inner: Tv_my_callable_1
    ) -> Tv_my_callable_1:
        return __my_callable_inner
    return my_decorator_inner(__my_callable_outer)
```

Output:
```
No configuration file found.
No pyproject.toml file found.
stubPath /Users/bb010g/python-typing-tests/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
/Users/bb010g/python-typing-tests/test_3212.py
  /Users/bb010g/python-typing-tests/test_3212.py:63:6 - error: Argument of type "(__my_callable_inner: Tv_my_callable_1@my_decorator_inner, /) -> Tv_my_callable_1@my_decorator_inner" cannot be assigned to parameter "__decorator" of type "MyCallableDecorator[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable, Tv_val_0@decorates_my_callable, Tv_val_1@decorates_my_callable, Tv_result@decorates_my_callable]" in function "decorates_my_callable"
    One or more overloads of "__call__" is not assignable
      Type "(__my_callable_inner: Tv_my_callable_1@my_decorator_inner, /) -> Tv_my_callable_1@my_decorator_inner" cannot be assigned to type "(__callable_0: My0Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]) -> My0Callable[Tv_val_0@decorates_my_callable, Tv_val_1@decorates_my_callable, Tv_result@decorates_my_callable]"
        Parameter 1: type "My0Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]" cannot be assigned to type "Tv_my_callable_1@my_decorator_inner"
          Type "My0Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]" cannot be assigned to type "MyCallable[object, object, object]"
        Function return type "Tv_my_callable_1@my_decorator_inner" is incompatible with type "My0Callable[Tv_val_0@decorates_my_callable, Tv_val_1@decorates_my_callable, Tv_result@decorates_my_callable]"
          Type "My0Callable[object, object, object]* | My1Callable[object, object, object]*" cannot be assigned to type "My0Callable[Tv_val_0@decorates_my_callable, Tv_val_1@decorates_my_callable, Tv_result@decorates_my_callable]"
            "My1Callable[object, object, object]*" is incompatible with protocol "My0Callable[Tv_val_0@decorates_my_callable, Tv_val_1@decorates_my_callable, Tv_result@decorates_my_callable]" (reportGeneralTypeIssues)
  /Users/bb010g/python-typing-tests/test_3212.py:68:31 - error: Argument of type "Tv_my_callable_0@my_decorator" cannot be assigned to parameter "__callable_1" of type "My1Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]" in function "__call__"
    Type "My0Callable[object, object, object]* | My1Callable[object, object, object]*" cannot be assigned to type "My1Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]"
      "My0Callable[object, object, object]*" is incompatible with protocol "My1Callable[Tv_val_0_co@decorates_my_callable, Tv_val_1_co@decorates_my_callable, Tv_result_co@decorates_my_callable]"
        "__call__" is an incompatible type
          Type "(tag: Literal[0], arg_0: object, arg_1: object) -> object" cannot be assigned to type "(tag: Literal[1], arg_0: Tv_val_1@My1Callable, arg_1: Tv_val_0@My1Callable) -> Tv_result@My1Callable"
            Parameter 1: type "Literal[1]" cannot be assigned to type "Literal[0]"
      TypeVar "Tv_result@My1Callable" is covariant
        Type "object" cannot be assigned to type "Tv_result_co@decorates_my_callable" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
Completed in 0.648sec
```

**VS Code extension or command-line**
Command line, version 1.1.230 from PyPI.

**Comment by erictraut** [comment]
Mypy also emits an error here, so I suspect that pyright is correct. I haven't dug into it because this is a very complex scenario. Are you able to repro the problem with a simpler example, or is this the minimal self-contained example that demonstrates the problem?

**Comment by erictraut** [comment]
Here's a simpler repro that type checks on mypy but fails on pyright:

```python
from typing import Callable, Protocol, TypeVar, Any

Tv_val_0 = TypeVar("Tv_val_0", contravariant=True)
Tv_val_0_co = TypeVar("Tv_val_0_co", covariant=True)
Tv_my_callable = TypeVar("Tv_my_callable", bound="MyCallable[object]")

class MyCallable(Protocol[Tv_val_0]):
    def __call__(self, __x: Tv_val_0) -> Any:
        ...

class MyDecorator(Protocol[Tv_val_0_co]):
    def __call__(self, __x: MyCallable[Tv_val_0_co]) -> Any:
        ...

def decorates_my_callable(__x: MyDecorator[Tv_val_0_co]) -> MyDecorator[Tv_val_0_co]:
    ...

def my_decorator_inner(__x: Tv_my_callable) -> Tv_my_callable:
    ...

decorates_my_callable(my_decorator_inner)
```

**Comment by erictraut** [comment]
I've fixed the bug that generated the first of the two errors. The fix for the first error will be included in the next release.

The second error (on the last return statement in the original sample) is legitimate, and mypy also emits an error for that line.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.231, which I just published. It will also be included in the next release of pylance.
